### PR TITLE
add Completions API support

### PR DESCRIPTION
**Pull request by mattf**
```
        Create a new NVIDIA LLM for Completions APIs.

        This class provides access to a NVIDIA NIM for completions. By default, it
        connects to a hosted NIM, but can be configured to connect to a local NIM
        using the `base_url` parameter. An API key is required to connect to the
        hosted NIM.

        Args:
            model (str): The model to use for reranking.
            nvidia_api_key (str): The API key to use for connecting to the hosted NIM.
            api_key (str): Alternative to nvidia_api_key.
            base_url (str): The base URL of the NIM to connect to.

        API Key:
        - The recommended way to provide the API key is through the `NVIDIA_API_KEY`
            environment variable.

        Additional arguments that can be passed to the Completions API:
        - max_tokens (int): The maximum number of tokens to generate.
        - stop (str or List[str]): The stop sequence to use for generating completions.
        - temperature (float): The temperature to use for generating completions.
        - top_p (float): The top-p value to use for generating completions.
        - frequency_penalty (float): The frequency penalty to apply to the completion.
        - presence_penalty (float): The presence penalty to apply to the completion.
        - seed (int): The seed to use for generating completions.
        - best_of (int): The number of completions to generate and return the best of.
        - echo (bool): Whether to echo the prompt in the completion.
        - logit_bias (Dict[str, float]): The logit bias to apply to the completion.
        - logprobs (int): The number of logprobs to return.
        - n (int): The number of completions to generate.
        - suffix (str): The suffix to use for generating completions.
        - user (str): The user ID to use for generating completions.

        These additional arguments can also be passed with `bind()`, e.g.
        `NVIDIA().bind(max_tokens=512)`, or pass directly to `invoke()` or `stream()`,
        e.g. `NVIDIA().invoke("prompt", max_tokens=512)`.
```

### CODE REVIEW

File: libs/ai-endpoints/docs/llms/nvidia_ai_endpoints.ipynb
Comment (dglogo) [inline libs/ai-endpoints/docs/llms/nvidia_ai_endpoints.ipynb:1]: lgtm!

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            "base_url",
+        ]:
+            if key in kwargs:
+                del kwargs[key]
+        self._init_args = self.__check_kwargs(kwargs)
+
+    @property
+    def available_models(self) -> List[Model]:
+        """
+        Get a list of available models that work with NVIDIARerank.
```
Comment (dglogo) [inline libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py:145]: NVIDIARerank -> NVIDIA

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        """
+        Create a new NVIDIA LLM for Completions APIs.
+
+        This class provides access to a NVIDIA NIM for completions. By default, it
+        connects to a hosted NIM, but can be configured to connect to a local NIM
+        using the `base_url` parameter. An API key is required to connect to the
+        hosted NIM.
+
+        Args:
+            model (str): The model to use for reranking.
```
Comment (dglogo) [inline libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py:89]: Completions

**Review by dglogo — APPROVED** [review]
Thanks for putting this together!

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    model: Optional[str] = Field(description="The model to use for completions.")
+
+    _base_url_var = "NVIDIA_BASE_URL"
+
+    _init_args: Dict[str, Any] = PrivateAttr()
+    """Stashed arguments given to the constructor that can be passed to
+    the Completions API endpoint."""
+
+    @root_validator(pre=True)
+    def _validate_base_url(cls, values: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (raspawar) [inline libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py:39]: we already have validation logic in client right? can this be moved to _common? I see repeating code in reranking and embedding class too

File: libs/ai-endpoints/tests/unit_tests/test_completions_models.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        func(llm, "IGNORED", arg_unknown="ARG")
+    assert len(record) == 1
+    assert "Unrecognized, ignored arguments: {'arg_unknown'}" in str(record[0].message)
+
+    bound_llm = llm.bind(bind_unknown="BIND")
+
+    with pytest.warns(UserWarning) as record:
+        func(bound_llm, "IGNORED")
+    assert len(record) == 1
+    assert "Unrecognized, ignored arguments: {'bind_unknown'}" in str(record[0].message)
```
Comment (raspawar) [inline libs/ai-endpoints/tests/unit_tests/test_completions_models.py:149]: Test cases look good!
Can we add for ainvoke, astream too?

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        if not payload.get("stream", True):
+            warnings.warn("stream set to false for streaming call, ignoring")
+            payload["stream"] = True
+
+        for chunk in self._client.get_req_stream(payload=payload):
+            content = chunk["content"]
+            generation = GenerationChunk(text=content)
+            if run_manager:  # todo: add tests for run_manager
+                run_manager.on_llm_new_token(content, chunk=generation)
+            yield generation
```
Comment (raspawar) [inline libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py:223]: Can we support _identifying_params: https://python.langchain.com/v0.1/docs/modules/model_io/llms/custom_llm/

File: libs/ai-endpoints/docs/llms/nvidia_ai_endpoints.ipynb [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.10.14"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
Comment (raspawar) [inline libs/ai-endpoints/docs/llms/nvidia_ai_endpoints.ipynb:222]: Can we add some examples, test cases for langchain chain, RunnableWithMessageHistory, RunnableLambda, RunnableParallel,  InMemoryCache, Output Parsers

This is some things I tried:
```python
from langchain_core.chat_history import InMemoryChatMessageHistory
from langchain_core.runnables.history import RunnableWithMessageHistory
from langchain_core.prompts import PromptTemplate

store = {} 
llm = NVIDIA().bind(max_tokens=256)

def get_session_history(session_id: str) -> InMemoryChatMessageHistory:
    if session_id not in store:
        store[session_id] = InMemoryChatMessageHistory()
    return store[session_id]

config = {"configurable": {"session_id": "1"}}

prompt = PromptTemplate.from_template("You're an assistant who's good at Writing Python Code with : {topic}. Respond in 20 words or fewer")
chain = prompt | llm
conversation = RunnableWithMessageHistory(
    chain,
    get_session_history,
)

print(conversation.invoke({"topic": "knapsack Problem"}, config=config))
```

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py (hunk 5 above)
Comment (mattf) [reply libs/ai-endpoints/langchain_nvidia_ai_endpoints/llm.py:223]: added

File: libs/ai-endpoints/docs/llms/nvidia_ai_endpoints.ipynb (hunk 6 above)
Comment (mattf) [reply libs/ai-endpoints/docs/llms/nvidia_ai_endpoints.ipynb:250]: let's follow up w/ this addition

File: libs/ai-endpoints/tests/unit_tests/test_completions_models.py (hunk 4 above)
Comment (mattf) [reply libs/ai-endpoints/tests/unit_tests/test_completions_models.py:149]: added

**Pull request merged**